pkg/appliance/upgrade_status.go

```diff
@@ -27,33 +27,37 @@ var defaultExponentialBackOff = &backoff.ExponentialBackOff{
 }
 
 type WaitForUpgradeStatus interface {
-	// WaitForUpgradeStatus does expodential backoff retries on upgrade status until it reaches a desiredStatuses and reports it to current <- string
+	// WaitForUpgradeStatus does exponential backoff retries on upgrade status until it reaches a desiredStatuses and reports it to current <- string
 	WaitForUpgradeStatus(ctx context.Context, appliance openapi.Appliance, desiredStatuses []string, undesiredStatuses []string, tracker *tui.Tracker) error
 }
 
 type UpgradeStatus struct {
 	Appliance *Appliance
 }
 
+type UpgradeStatusCtx string
+
+const UpgradeStatusGetErrorMessage UpgradeStatusCtx = "UpgradeStatusGetErrorMessage"
+
 func (u *UpgradeStatus) upgradeStatus(ctx context.Context, appliance openapi.Appliance, desiredStatuses []string, undesiredStatuses []string, tracker *tui.Tracker) backoff.Operation {
 	name := appliance.GetName()
 	logEntry := log.WithField("appliance", name)
 	logEntry.WithField("want", desiredStatuses).Info("polling for upgrade status")
-	hasRebooted := false
 	return func() error {
 		ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
 		defer cancel()
 		status, err := u.Appliance.UpgradeStatus(ctx, appliance.GetId())
 		if err != nil {
 			if tracker != nil {
-				if errors.Is(err, context.DeadlineExceeded) {
-					tracker.Update("rebooting, waiting for appliance to come back online")
-					hasRebooted = true
-				} else if hasRebooted {
-					tracker.Update("switching partition")
-				} else {
-					tracker.Update("installing")
+				msg := "no response, appliance offline"
+				if !errors.Is(err, context.DeadlineExceeded) {
+					msg = err.Error()
 				}
+				if v, ok := ctx.Value(UpgradeStatusGetErrorMessage).(string); ok {
+					msg = v
+				}
+
+				tracker.Update(msg)
```

Comment:
I don't think using the context to pass messages is a good idea in this case. For starters, the "no response, appliance offline" message will never be shown to the user because it is immediately replaced by the error passed in context on line 57, so L52-L55 is effectively dead code.
But more importantly, we lose a lot of fine grained control over messages here. The rebooting logic that is removed here was there because it was once asked for, and it gives the user a lot more information on what's going on even when the API is unreachable. This change just makes it less informative and reverts a previous feature request, but I don't really understand why this is a necessary change? What does this fix?

Reply:
> For starters, the "no response, appliance offline" message will never be shown to the user because it is immediately

No this is not true.
 See image for example.
![img-2022-11-15-110914](https://user-images.githubusercontent.com/1218385/201893150-f62b5def-1e75-432d-8dc4-d2258e4079f4.png)
In this case, the appliance got terminated mid operation.

 Upgrade status is viewed both in prepare and complete. it all depends on ctx and what type of error we get.

> But more importantly, we lose a lot of fine grained control over messages here.

I would argue we get even more control with this implementation then the prior. Now we can acctually append correct error message, depending on when the function is being used.


> What does this fix?

It no longer report false error messages, if a appliance is offline/internal error, it would report as "rebooting", but that would not always be true.

Reply:
> I would argue we get even more control with this implementation then the prior. Now we can acctually append correct error message, depending on when the function is being used.

So how does this differentiate between a 504 response and no response at all? How do you determine if a reboot has happened or not and present different information depending on that. Like I mentioned, that was a feature that was implemented on request. We can still use the values in context, but that doesn't really explain why the reboot logic was removed. They are not mutually excluding each other.

> It no longer report false error messages, if a appliance is offline/internal error, it would report as "rebooting", but that would not always be true.

That could have been solved in other ways. It should not show "rebooting" on prepare, but on complete when it is actually rebooting. That's a simple boolean switch.